processing/src/main/java/io/druid/segment/incremental/OffheapIncrementalIndex.java

```diff
@@ -71,37 +84,61 @@
 
   private String outOfRowsReason = null;
 
+  private final File tmpDirLocation;
+
   public OffheapIncrementalIndex(
       IncrementalIndexSchema incrementalIndexSchema,
       boolean deserializeComplexMetrics,
       boolean reportParseExceptions,
       boolean sortFacts,
-      int maxRowCount,
-      StupidPool<ByteBuffer> bufferPool
+      long maxOffheapSize
   )
   {
     super(incrementalIndexSchema, deserializeComplexMetrics, reportParseExceptions, sortFacts);
-    this.maxRowCount = maxRowCount;
-    this.bufferPool = bufferPool;
+    tmpDirLocation = Files.createTempDir();
+
+    try {
+      File mapDbFile = File.createTempFile("druid-offheap-index-mapdb", "", tmpDirLocation);
+      File aggsBBFile = File.createTempFile("druid-offheap-index-aggs", "", tmpDirLocation);
+
+      AtomicLong availableTotalSpace = new AtomicLong(maxOffheapSize);
+
+      final DBMaker dbMaker = new MyDBMaker(mapDbFile, 24, 29, availableTotalSpace)
+          .deleteFilesAfterClose()
+          .commitFileSyncDisable()
+          .cacheSize(1024)
+          .transactionDisable()
+          .mmapFileEnable();
+
+      this.factsDb = dbMaker.make();
+
+      if (sortFacts) {
+        TimeAndDimsBTreeKeySerializer timeAndDimsSerializer = new TimeAndDimsBTreeKeySerializer(this);
+        this.facts = factsDb.createTreeMap("__facts" + UUID.randomUUID())
+                            .keySerializer(timeAndDimsSerializer)
+                            .comparator(timeAndDimsSerializer.getComparator())
+                            .valueSerializer(Serializer.INTEGER)
+                            .make();
+      } else {
+        this.facts = factsDb.createHashMap("__facts" + UUID.randomUUID())
+                            .keySerializer(TimeAndDimsSerializer.INSTANCE)
+                            .valueSerializer(Serializer.INTEGER)
+                            .make();
+      }
 
-    if (sortFacts) {
-      this.facts = new ConcurrentSkipListMap<>(dimsComparator());
-    } else {
-      this.facts = new ConcurrentHashMap<>();
-    }
+      this.aggBBFactory = new DiskBBFactory(aggsBBFile, 29, availableTotalSpace);
```

Comment:
no, both aggregation buffer and mapdb buffers are being allocated from same "availableTotalSpace" (its an AtomicLong and not long). so both keep on incrementing it everytime they allocate a buffer.